pkg/infrastructure/clusterapi/clusterapi.go

```diff
@@ -111,8 +120,11 @@ func (i InfraProvider) Provision(dir string, parents asset.Parents) ([]*asset.Fi
 		ClusterID:     clusterID.InfraID,
 		InstallConfig: installConfig,
 	}
-	if err := i.capiProvider.PreProvision(preProvisionInput); err != nil {
-		return fileList, fmt.Errorf("failed during pre-provisioning: %w", err)
+
+	if p, ok := i.capiProvider.(PreProvider); ok {
```

Comment:
@pierreprinetti @andfasano @joelspeed is this more in line with the interface pattern suggested/discussed that you had in mind? i broke PreProvision into its own interface, test whether the passed-in `capiProvider` implements the interface, if so call the function. 

i can flesh this out a bit more, but so far so good, just wanted to make sure this is what you had in mind.

Reply:
Thanks @patrickdillon that sounds fine to me, at least from my point of view it was the approach I was referring to during the discussion.

For the name bikeshedding, it could be useful to have a common prefix shared by all of the interfaces, ie `CAPILifecyclePreProvider`, just to highlight the common trait of the interfaces (being an hook of the CAPI life cycle), but so for far it's fine.

Reply:
Yeah this is the kind of think I was imagining, so that we can just implement the specific sub interfaces that we need rather than implementing the whole thing when not needed, thanks

Reply:
Done. Thanks!

Reply:
Yes this is what I had in mind, thank you.

Note that this "interface assertion" pattern means that a platform method with the wrong signature, or the wrong name, is silently ignored. As a mitigation, I found it useful to [validate adherence](https://github.com/shiftstack/installer/blob/a18f2c1c40a79afa2d7cfb778e7851d427cc23d0/pkg/infrastructure/openstack/clusterapi/clusterapi.go#L40-L42) to the interface with this compile-time declaration:

```Go
var _ clusterapi.PreProvider = Provider{}
func (p Provider) PreProvision(ctx context.Context, in clusterapi.PreProvisionInput) error {
```